### PR TITLE
Add partial support for second-order derivatives (for grid.h's input)

### PR DESCRIPTION
**Pull request by ventusff**
Hi,

As discussed in issue https://github.com/NVlabs/tiny-cuda-nn/issues/58, a backward_backward functionality is needed.

For now, I managed to add a partial support for second-order derivatives, **only** for `grid.h`, and **only** for `d(dL_dinput)_d(...)` . 
This satifies the need of backwarding nablas's gradients toward grid params and downstream modules' params.

**edit: now support d(dL_dinput)_d(input)**
```
# NOTE: currently support:
#  ✓ d(dL_dinput)_d(dL_doutput)  ->  kernel_grid_backward_input_backward_dLdoutput
#  ✓ d(dL_dinput)_d(params)      ->  kernel_grid_backward_input_backward_grid
#  ✓ d(dL_dinput)_d(input)       ->  kernel_grid_backward_input_backward_input
#  x d(dL_dparam)_d(...)
```

I tried my best to follow the code style of your original project, and have done various tests to ensure that the code is running correctly. 
The changes are made as little as I can, and the compiling has passed with no errors.

## Code tests
I provide three testing tools for the newly added `backward_backward_input` functionality:
https://gist.github.com/ventusff/57f47588eaff5f8b77a382260e7da8a3 

1. :heavy_check_mark:  test_train(): train a toy SDF model with eikonal term.
2. :heavy_check_mark:  grad_check(): check backward_backward **numerical correctness** via `torch.autograd.gradcheck`.
3. :heavy_check_mark:  vis_graph(): visualize torch compute graph

## Toy model compute graph:
toy_model:
```python
class SDF(nn.Module):
    def __init__(self, hash=True, n_levels=12, log2_hashmap_size=15, base_resolution=16) -> None:
        super().__init__()
        self.encoder = tcnn.Encoding(3, {
            "otype": "HashGrid" if hash else "DenseGrid",
            "n_levels": n_levels,
            "n_features_per_level": 2,
            "log2_hashmap_size": log2_hashmap_size,
            "base_resolution": base_resolution,
            "per_level_scale": 1.5
        })
        self.decoder = nn.Sequential(
            nn.Linear(self.encoder.n_output_dims, 64),
            nn.ReLU(True),
            nn.Linear(64, 1)
        )
    
    def forward(self, x):
        encoded = self.encoder(x).to(dtype=torch.float)
        sdf = self.decoder(encoded)
        return sdf
    
    def forward_with_nablas(self, x):
        with torch.enable_grad():
            x = x.requires_grad_(True)
            sdf = self.forward(x)
            nablas = autograd.grad(
                sdf,
                x,
                torch.ones_like(sdf, device=x.device),
                create_graph=True,
                retain_graph=True,
                only_inputs=True)[0]
        return sdf, nablas
```

You can see that the gradients of `nablas` are backwared towards `grid_params` (needs `d(dL_dx)_dgrid)`) and `decoder.xxx` (needs `d(dL_dx)_d(dL_doutput)`) through `backwardBackward`.
attached

## Theoretical derivation
From gradients of `dL_dx`,  to gradients of `dL_dgrid` and `dL_d(dL_doutput)`.
second-order-grad

### **edit: add theoretical derivations of d(dy_dx)_dx**
image

~~1. Since I add a **virtual** function `backward_backward_input` in `object.h:DifferentiableObject`, I must put a dummy error function in `network.h:Network` and `encoding.h:Encoding` to pass compiling, which inherit from `DifferentiableObject`. 
Currently this funciton just throw a NotImplementedError. I hope this won't hurt any current mechanisms, and you can finish the remain support for backward_backward later to cancel this dummy function.~~

### CODE REVIEW

File: bindings/torch/tinycudann/modules.py [hunk 1]
```diff
@@ -55,6 +55,79 @@ def backward(ctx, doutput):
 
 		return None, input_grad, weight_grad, None
 
+# NOTE: Jianfei: this is a temporary solution, when only grid has backward_backward and networks do not.
```
Comment (Tom94) [inline bindings/torch/tinycudann/modules.py:58]: I would prefer if all modules shared the same `_module_function` and `_module_function_backward` in the interest of avoiding code duplication. Since their interface is equal, this should be possible without overhear, right? (Except that those components not implementing double-bwd will throw an exception.)

File: bindings/torch/tinycudann/modules.py [hunk 2]
```diff
@@ -55,6 +55,79 @@ def backward(ctx, doutput):
...
+			input_grad, weight_grad = ctx_fwd.native_tcnn_module.bwd(ctx_fwd.native_ctx, input, params, output, scaled_grad)
+			input_grad = None if input_grad is None else (input_grad / ctx_fwd.loss_scale)
+			weight_grad = None if weight_grad is None else (weight_grad / ctx_fwd.loss_scale)
+		return input_grad, weight_grad
+
+	@staticmethod
+	def backward(ctx, dinput_grad, dweight_grad):
+		# NOTE: currently support:
+		#		✓	d(dL_dinput)_d(dL_doutput)	doutput_grad
+		#		✓	d(dL_dinput)_d(params)		weight_grad
```
Comment (Tom94) [inline bindings/torch/tinycudann/modules.py:104]: Minor nitpick: tabs should only be used for indentation (i.e. in the beginning of lines), not for alignment.

File: bindings/torch/tinycudann/modules.py [hunk 3]
```diff
@@ -55,6 +55,79 @@ def backward(ctx, doutput):
...
+		#		x	d(dL_dinput)_d(input)
+		#		x	d(dL_dparam)_d(...)
+		# assert dweight_grad is None, "currently do not support 2nd-order gradients from gradient of grid"
+		input, params, doutput = ctx.saved_tensors
+		with torch.enable_grad():
+			# NOTE: preserves requires_grad info
+			# NOTE: be cautious when multiplying and dividing loss_scale:
+   			# dinput_grad = dinput_grad * ctx.ctx_fwd.loss_scale
+			doutput = doutput * ctx.ctx_fwd.loss_scale
```
Comment (Tom94) [inline bindings/torch/tinycudann/modules.py:113]: Is it important to do this in `with torch.enable_grad():` rather than just outside of `with torch.no_grad():`? Or in other words: if `doutput` doesn't require gradients in the first place, I don't think they need to be explicitly enabled.

File: bindings/torch/tinycudann/modules.py [hunk 4]
```diff
@@ -217,3 +290,21 @@ def __init__(self, n_input_dims, encoding_config, seed=1337, dtype=None):
 
 	def _native_tcnn_module(self):
 		return _C.create_encoding(self.n_input_dims, self.encoding_config, self.precision)
+
+	def forward(self, x):
```
Comment (Tom94) [inline bindings/torch/tinycudann/modules.py:294]: This can be removed if `_module_function` is unified.

File: include/tiny-cuda-nn/encoding.h [hunk 5]
```diff
@@ -72,6 +72,17 @@ class Encoding : public DifferentiableObject<float, T, T> {
 	size_t n_params() const override { return 0; }
 
 	std::vector<std::pair<uint32_t, uint32_t>> layer_sizes() const override { return {}; }
+
+	void backward_backward_input(
```
Comment (Tom94) [inline include/tiny-cuda-nn/encoding.h:76]: I'd prefer if `DifferentiableObject::backward_backward_input` already threw the exception. That'd save the trouble of overriding it in `Network` and `Encoding`.

File: include/tiny-cuda-nn/encodings/grid.h [hunk 6]
```diff
@@ -485,6 +485,161 @@ __global__ void kernel_grid_backward_input(
...
+	for (uint32_t f = 0; f < N_FEATURES_PER_THREAD; ++f) {
+		grad[f] = dL_dy[i + (level * N_FEATURES_PER_LEVEL + feature + f) * num_elements];
+	}
+
+	if (interpolation_type == InterpolationType::Nearest) {
+		// d(dydx)_dgrid is zero when there's no interpolation.
+		return;
+	}
+
+	// for N-linear interpolation
```
Comment (Tom94) [inline include/tiny-cuda-nn/encodings/grid.h:581]: I think it's important to also support `interpolation_type == InterpolationType::Smoothstep`, which should amount to a simple additional factor (by the chain rule). Analogous to how dL_dinput is currently computed.

`pos_fract` would require an additional overload that also computes second derivatives of the smoothstep, which are simple polynomials.

**Review by Tom94 — COMMENTED** [review]
Hi Jianfei, thank you very much for this PR -- I think this is great functionality to have, even if implemented just for the encoding!

This is also a great implementation, I have very little to criticize. Here are just a few small requests before I feel comfortable merging.

**Comment by Tom94** [comment]
Also just wanted to say that I _very much_ appreciate the detailed derivations and explanations in the PR. This is really something to point other people to as an example of how it should be done!

File: bindings/torch/tinycudann/modules.py (hunk 3 above)
Comment (ventusff) [reply bindings/torch/tinycudann/modules.py:113]: Yes, I found it necessary.

1. Shortly speaking, the backward.backward() function here is invoked implicitly in an `no_grad()` context by default. Hence, the product of multiplying a `requires_grad=True` Tensor (doutput) with a float (loss_scale), will be a `requires_grad=False` Tensor by default. The `requires_grad` property is lost by default.

Specifically speaking, consider the final `loss.backward()`, which typically will _NOT_ specify `create_graph=True` (and usually not any other parameters at all). Then, it will invoke all tensors `backwards` via a [run_backward()](https://github.com/pytorch/pytorch/blob/e813eaee22be2bfe5e6f2a137305cffe864a01da/torch/autograd/__init__.py#L173) funtion, which causing all the backwards() implicitly running in a `create_graph=False / no_grad()` context, including the one here. This introduces the above mentioned problem of not passing `requires_grad` property.

2. The underlying `bwd_bwd_input` in `bindings.cpp` relies on the `requires_grad()` property of `doutput` to behave corretly (to determine whether to compute `dL_ddLdoutput` or not). 

I've thought of other alternative ways, e.g. passing a bool `need_xxx_grad` from python to cpp,  or multiplying the `loss_scale` inside `bindings.cpp` instead of in `modules.py`. But, they may not be as elegant ? or as concise? comparing to the implementation now. So `requires_grad` property indeed matters.

Comment (Tom94) [reply bindings/torch/tinycudann/modules.py:113]: I see, that makes perfect sense -- I hadn't realized that the entire function is implicitly wrapped in `no_grad()`. Thank you for the clarification :)

File: include/tiny-cuda-nn/encodings/grid.h (hunk 6 above)
Comment (ventusff) [reply include/tiny-cuda-nn/encodings/grid.h:581]: It is computing `d(dy_dx)_dparams` here, which only needs first-order derivatives of `identity_fun` and `smoothstep`, as derived below:
(mainly because we are deriving w.r.t. grid params now)
image

This part is already taken care of, by 
```c++
// ...
pos_fract(positions_in[i + dim * num_elements], &pos[dim], &pos_derivative[dim], &pos_grid[dim], scale, smoothstep, smoothstep_derivative);
// pos_derivative is multiplied
float grad_in = scale * ((float*)dL_ddLdx(i))[grad_dim] * pos_derivative[grad_dim];
```

Comment (ventusff) [reply include/tiny-cuda-nn/encodings/grid.h:581]: I think you were saying `d(dy_dx)_dx`. 
This is currently not supported, only because in my use case we don't consider the eikonal term's influence on input x itself. 

I can add this support though, which requires another kernel function. It will be a little messy, because `d(doutput_dx,y,z)_d(x) = p(doutput_dx)_px (=0) + p(doutput_dy)_px + p(doutput_dz)_px`.

File: include/tiny-cuda-nn/encodings/grid.h [hunk 7]
```diff
@@ -485,6 +485,161 @@ __global__ void kernel_grid_backward_input(
...
+    #pragma unroll
+    for (uint32_t grad_dim = 0; grad_dim < N_POS_DIMS; ++grad_dim) {
+        float grad_in = scale * ((float*)dL_ddLdx(i))[grad_dim] * pos_derivative[grad_dim];;
```
Comment (ventusff) [inline include/tiny-cuda-nn/encodings/grid.h:584]: By multiplying `pos_derivative`, `smoothstep`'s contribution in `d(dy_dx)_dparams` is taken care of.
(oops, the extra semicolon is a typo. fixed soon.)

File: include/tiny-cuda-nn/encodings/grid.h (hunk 6 above)
Comment (ventusff) [reply include/tiny-cuda-nn/encodings/grid.h:581]: image

File: bindings/torch/tinycudann/modules.py [hunk 8]
```diff
@@ -37,7 +37,6 @@ def forward(ctx, native_tcnn_module, input, params, loss_scale):
 		return output
 
 	@staticmethod
-	@once_differentiable
```
Comment (ventusff) [inline bindings/torch/tinycudann/modules.py:40]: I'm not sure why there's once_differentiable here --- but double backwards typically requires multiple calling to backward(). 

For example, first compute nablas needs one call, then compute nablas' loss backward needs another call.

Deleting it in my case works well & computes correctly, and I think most of the modules in the repo actually support multiple time backwards calling, since the params and gradients pointers are set everytime.

**Comment by ventusff** [comment]
Thanks for the nice tips and kind words :) I resolved some of the suggestions now. There are a few remains, mostly about computing that I'm not sure.

**Comment by ventusff** [comment]
Hi @Tom94,
I've finished the following implementation and tests, and I think it's probably ready to merge now :smiley: 
- [x] update with your latest convention of code and data convention, inlucding `MatrixView` and `xxx_impl`
- [x] as you suggested, I add an implementation of `d(dL_dx)_dx`, which requires: 
  - [x] an implementation of  `d(dy_dx)_dx`
  - [x] an overload of `pos_fract` for second-order derivatives, and second-order derivatives for `Smoothstep`
  - [x] an additional kernel `kernel_grid_backward_input_backward_input`
  - [x] bindings update and modules udpate
- [x] passed additional `gradcheck` and `gradgradcheck` for the newly added `d(dL_dx)_dx`'s calculation; check numerical correctness under both `Linear` interpolation and `Smoothstep` interpolation (the previous gist script for testing is also updated)

I will clean and upload theoretical derivations for `d(dy_dx)_dx` soon. In the meantime you can review my changes :)

**Comment by Tom94** [comment]
Wow, this is incredible, thank you so much for adding `d(dL_dx)_dx`! I haven't had the time to get back to this PR during the week and did not expect the follow-up. Will go through it now. :)

By the way: do you want to add your test script to the `scripts` folder? It would surely come in handy in the future to ensure nothing regresses. (E.g. when support for non-contiguous inputs from torch is added.)

File: bindings/torch/tinycudann/modules.py (hunk 8 above)
Comment (Tom94) [reply bindings/torch/tinycudann/modules.py:40]: Makes perfect sense. `@once_differentiable` was used because none of the components of tcnn supported double-backward until this PR came along.

File: include/tiny-cuda-nn/encodings/grid.h [hunk 9]
```diff
@@ -782,6 +1101,133 @@ class GridEncodingTemplated : public GridEncoding<T> {
...
+			// from dL_d(dL_dx) to dL_doutput
+			linear_kernel(kernel_grid_backward_input_backward_dLdoutput<T, N_POS_DIMS>, 0, stream, 
+				num_elements,
+				m_n_features,
+				// inputs
+				dL_ddLdinput.view(),
+				forward.dy_dx.data(),
+				dL_dy_rm,
+				// outputs
+				dL_ddLdoutput->pitched_ptr()
```
Comment (Tom94) [inline include/tiny-cuda-nn/encodings/grid.h:1198]: Could you replace this pitched pointer with `->view()` in order to also support row-major `dL_ddLdoutput`?

**Comment by Tom94** [comment]
Turns out that was the only nitpick I could find in the code -- so I quickly made the change myself and merged.

Thanks again for adding all this. I can't stress enough how much I appreciate this sort of high-quality code contribution!

**Comment by juuso-oskari** [comment]
@ventusff Thank you so much for your work on this. Are you still planning to do the double backwards for the fully_fused_mlp.cu? I would really like to test it on my thesis project.
